contracts/BC_fusion/StakePool.sol

```diff
@@ -0,0 +1,290 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/CountersUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/structs/DoubleEndedQueueUpgradeable.sol";
+
+import "./System.sol";
+
+interface IStakeHub {
+    function unbondPeriod() external view returns (uint256);
+    function transferGasLimit() external view returns (uint256);
+    function poolImplementation() external view returns (address);
+}
+
+contract StakePool is Initializable, ReentrancyGuardUpgradeable, ERC20Upgradeable, System {
+    using CountersUpgradeable for CountersUpgradeable.Counter;
+    using DoubleEndedQueueUpgradeable for DoubleEndedQueueUpgradeable.Bytes32Deque;
+
+    /*----------------- constant -----------------*/
+    uint256 public constant COMMISSION_RATE_BASE = 10_000; // 100%
+
+    /*----------------- storage -----------------*/
+    address public validator; // validator's operator address
+    uint256 public totalPooledBNB; // total reward plus total BNB staked in the pool
+
+    // hash of the unbond request => unbond request
+    mapping(bytes32 => UnbondRequest) private _unbondRequests;
+    // user => unbond request queue(hash of the request)
+    mapping(address => DoubleEndedQueueUpgradeable.Bytes32Deque) private _unbondRequestsQueue;
+    // user => locked shares
+    mapping(address => uint256) private _lockedShares;
+    // user => personal unbond sequence
+    mapping(address => CountersUpgradeable.Counter) private _unbondSequence;
+
+    // for slash
+    bool private _freeze;
+    uint256 public remainingSlashBnbAmount;
```

Comment:
I think we don't have `remainingSlashBnbAmount` any longer.

Because we ensure that a validator can not undelegate to a threshold, otherwise it will be jailed.

So we can ensure that it always have some delegation above a threshold if it is active.  We just simplify the design: if there is enough delegation for the slash, then do it; if not enough, just slash all; no need to record the `remainingSlashBnbAmount`.

Reply:
Done